### PR TITLE
New Home

### LINKED ISSUE #173: Port the home page to match mockups

### PR DESCRIPTION
**Pull request by deepansh96**
Fixes #173 

## Summary
- [x] Replaced the `copy.svg` that is used alongside the URL component. Reason - the existing SVG had width and heigh predefined inside and hence made scaling according to its parent component messy. Configured it such that it will always scale to its parent component without having to specify height or width
- [x] Added 2 new svgs that will be used in the table column headers - to depict sorting
- [x] Some style changes in `SimpleBadge.vue`. The padding was hardcoded by default, made that padding configurable. This was done because badge was needed to be shown in the table and the size had to be configured
- [x] Added a new component `PlioListItem`. This component contains the details of a plio viz - date updated, status, plio title, plio URL and some action buttons (play, edit, duplicate). It takes a plioID as a prop and then fetches the respective plio's data
- [x] Added a rough skeleton for `PlioListItem`
- [x] Added a `Table` component. Takes in two props - `data` and `columns`. Both are arrays. If `columns` are `["col1", "col2", "col3"]` then the `data` prop will have a format like 
```
data = [
	{
		"col1" : {},
		"col2" : "qwerty",
		"col3" : "30.5"
	},

	{
		"col1" : "50 %",
		"col2" : "xyz",
		"col3" : {
			"type" : "component",
			"value": "asdqiweqkj"
		}
	}
]
```
This means that a cell of the table can be anything, even a component
- [x] Changed `URL.vue` to make the underlining configurable by props. It was defaulted before.
- [x] Added a `NewHome.vue` page - fetches the uuids of all plios -> prepares the data arrays to pass to the table

## Test Plan
Tested locally

TODO
- [x] Searching/Filtering in the table
- [x] Sorting the columns in the table
- [x] Mobile responsive table

### CODE REVIEW

File: src/components/UI/Table/Table.vue [hunk 1]
```diff
@@ -25,21 +25,41 @@
...
+                      {{ capitalize(columnName) }}
+                    </div>
+                    <div class="p-1 my-auto">
+                      <inline-svg
+                        :src="require('@/assets/images/chevron-down-solid.svg')"
+                        class="h-3 w-3 my-1"
+                      ></inline-svg>
+                    </div>
+                  </div>
+                </th>
```
Comment (deepansh96) [inline src/components/UI/Table/Table.vue:54]: Will explain this component on call - yet to add comments and structure it in a better way

File: src/services/API/Plio.js [hunk 2]
```diff
@@ -40,9 +40,25 @@ export default {
...
-    return apiClient().get(pliosEndpoint);
+  getAllPlios(userId = null) {
+    // returns all the plios created by the user if userId is present
+    // else returns all the plios
+    if (userId == null) return apiClient().get(pliosEndpoint);
+    else
+      return apiClient().get(pliosEndpoint, {
+        params: { user: userId },
+      });
+  },
```
Comment (dalmia) [inline src/services/API/Plio.js:51]: don't need to pass user Id - will always return plios of the logged in user.

File: src/services/API/Plio.js [hunk 3]
```diff
@@ -40,9 +40,25 @@ export default {
...
+
+  getAllPlioIds(userId = null) {
+    // returns a flat list of plio uuids
+    // created by userId (if available)
+    var url = pliosEndpoint + "list_uuid";
+    if (userId == null) return apiClient().get(url);
+    else
+      return apiClient().get(url, {
+        params: { user: userId },
+      });
```
Comment (dalmia) [inline src/services/API/Plio.js:61]: same comment as above.

File: src/components/UI/Skeletons/PlioListItemSkeleton.vue [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<template>
```
Comment (dalmia) [inline src/components/UI/Skeletons/PlioListItemSkeleton.vue:1]: guessing that needs to be changed.

**Comment by dalmia** [comment]
- web version: 
   - need to show a number of all the Plios as shown in the mockup
   - the name table should be fixed size - currently goes to max size - column names should be in one line
   - in the name - after one limit ....
   - search bar right
   - search bar thinner
   - search functionality
   - sort functionality
   - weird placeholder skeleton
   - workspace switcher not working
   - "untitled" for empty plio

- mobile version
   - left/right padding reduce
   - space needs to be optimized
   - text needs to be reduced
   -  remove details etc when in mobile mode
   - reduce things etc slowly

File: src/components/UI/ListItems/PlioListItem.vue [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+      },
+      duplicateButtonClass: "bg-gray-100 hover:bg-gray-200 rounded-md shadow-md h-10",
+      urlStyleClass: "text-xs font-bold text-yellow-600",
+      urlCopyButtonClass: "text-yellow-600",
+      isDataLoaded: false,
+    };
+  },
+
+  async created() {
+    this.isDataLoaded = false;
```
Comment (dalmia) [inline src/components/UI/ListItems/PlioListItem.vue:122]: why is this needed?

File: src/components/UI/ListItems/PlioListItem.vue [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        "text-green-500 border-green-500": this.isPublished,
+        "border-black text-black": !this.isPublished,
+        "text-xs": true,
+        "px-2 py-1": true,
+      };
+      return badgeClass;
+    },
+    updatedAt() {
+      var humanReadableDate = new Date(this.plioDetails.updated_at).toDateString();
+      return humanReadableDate;
```
Comment (dalmia) [inline src/components/UI/ListItems/PlioListItem.vue:151]:
```suggestion
      return new Date(this.plioDetails.updated_at).toDateString();
```

File: src/components/UI/ListItems/PlioListItem.vue [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    },
+    title() {
+      return this.plioDetails.plioTitle;
+    },
+    plioLink() {
+      // prepare the link for the plio from the plio ID
+      if (this.plioId == "") {
+        return "";
+      }
+      return process.env.VUE_APP_FRONTEND + "/#/play/" + this.plioId;
```
Comment (dalmia) [inline src/components/UI/ListItems/PlioListItem.vue:164]: this will have org URL for plios created in org workspace

File: src/components/UI/ListItems/PlioListItem.vue [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        this.plioDetails = plioDetails;
+        this.$emit("fetched", this.plioDetails);
+      });
+    },
+    playPlio() {
+      this.$router.push({ name: "Player", params: { plioId: this.plioId } });
+    },
+    editPlio() {
+      this.$router.push({ name: "Editor", params: { plioId: this.plioId } });
+    },
```
Comment (dalmia) [inline src/components/UI/ListItems/PlioListItem.vue:181]: need to pass org as param as well

**Review by dalmia — CHANGES_REQUESTED** [review]
Didn't review NewHome and Table

File: src/components/UI/Table/Table.vue [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+      //     });
+      //   });
+      // }
+      // if (sortKey) {
+      //   data = data.slice().sort(function (a, b) {
+      //     a = a[sortKey];
+      //     b = b[sortKey];
+      //     return (a === b ? 0 : a > b ? 1 : -1) * order;
+      //   });
+      // }
```
Comment (rathorevaibhav) [inline src/components/UI/Table/Table.vue:175]: are these needed for later?

File: src/services/API/Session.js [hunk 10]
```diff
@@ -34,4 +34,11 @@ export default {
       sessionAnswerDetails
     );
   },
+
+  getUniqueUsersCount(plioId) {
+    // get the count of unique users who watched the given plio
+    return apiClient().get(sessionsEndpoint + "unique_users", {
+      params: { plio: plioId },
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/services/API/Session.js:41]: In the backend, the plio Id is being used in the queryset of the session and the endpoint for unique_users uses that queryset to do further processing. I agree though that the right place for that function seems to be in `plio/views.py`

File: src/services/API/Plio.js [hunk 11]
```diff
@@ -40,9 +40,25 @@ export default {
...
+    // returns all the plios created by the user if userId is present
+    // else returns all the plios
+    if (userId == null) return apiClient().get(pliosEndpoint);
+    else
+      return apiClient().get(pliosEndpoint, {
+        params: { user: userId },
+      });
+  },
+
+  getAllPlioIds(userId = null) {
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/services/API/Plio.js:53]: I think the point here was to avoid fetching all the data when we only need the UUID. Maybe we can say pass a param in the `getAllPlios` API call to just return the given set of keys. @deepansh96 ?

File: src/components/UI/ListItems/PlioListItem.vue (hunk 5 above)
Comment (deepansh96) [reply src/components/UI/ListItems/PlioListItem.vue:132]: Added it so that I can show the skeleton until a plio's data is loaded
- I think it's better to do it using vuex ka startLoading and stopLoading

File: src/components/UI/Skeletons/PlioListItemSkeleton.vue (hunk 4 above)
Comment (deepansh96) [reply src/components/UI/Skeletons/PlioListItemSkeleton.vue:1]: Yes will edit this, looks very weird right now

File: src/components/UI/Table/Table.vue (hunk 9 above)
Comment (deepansh96) [reply src/components/UI/Table/Table.vue:175]: yes, to implement the sorting and searching logic

**Comment by deepansh96** [comment]
- web version:
	- ✅ need to show a number of all the Plios as shown in the mockup
	- ✅ ~the name table should be fixed size - currently goes to max size~ -> deferred for now
	- ✅ column names should be in one line
	- ✅ in the name - after one limit .... (ellipsis truncate)
	- ✅ search bar right
	- ✅ search bar thinner
	- ✅ search functionality
	- ✅ sort functionality
	- ✅ weird placeholder skeleton - plioListItem - fixed placeholder skeleton
	- ✅ "untitled" for empty plio
	- ✅ duplicate functionality
	- ✅ remove the hardcoded data values
	- ✅ hover button pointer on sorting columns

  - mobile version
	- ✅ left/right padding reduce
	- ✅ space needs to be optimized
	- ✅ text needs to be reduced
	- ✅ remove details etc when in mobile mode
	- ✅ reduce things etc slowly


TODO:
- [ ] workspace switcher not working
- [ ] test multitenancy

**Comment by dalmia** [comment]
- when searching, it seems to reload all the plios (notice the attachment - screen opacity changes - which I assumed means that it is fetching again from the server?)
- remove old loading spinner
- placeholders come a few seconds after page has been loaded - should come instantly
- need to add a placeholder or something for when there are no plios created
- apart from the sorting based on name or number of learners as the first level of sorting, the second level of sorting should be based on time 
- workspace switcher not working - plios loaded of older plios instead of the workspace



https://user-images.githubusercontent.com/13185699/115135708-1c662100-a038-11eb-870b-922d8ea8c4c9.mov

File: src/components/UI/Table/Table.vue [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+              return Object.keys(objectToFilter).some((value) => {
+                return (
+                  String(objectToFilter[value]).toLowerCase().indexOf(filterKey) > -1
+                );
+              });
+            } else return String(objectToFilter).toLowerCase().indexOf(filterKey) > -1;
+          });
+        });
+      }
+      return data;
```
Comment (dalmia) [inline src/components/UI/Table/Table.vue:238]: has this been taken from somewhere?

File: src/components/UI/Table/Table.vue [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+      this.sortOrders[key] = this.sortOrders[key] * -1;
+    },
+    savePlioDetails(entryIndex, plioDetails) {
+      // save the plio details after they are fetched from the PlioListItem
+      // this is done so that the searching/sorting can be done on the plios as well
+      if (this.filteredData != undefined && this.filteredData[entryIndex] != undefined) {
+        this.filteredData[entryIndex]["name"] = {
+          ...this.filteredData[entryIndex]["name"],
+          ...plioDetails,
+        };
```
Comment (dalmia) [inline src/components/UI/Table/Table.vue:219]: didn't understand what this means

**Review by dalmia — CHANGES_REQUESTED** [review]

File: src/components/UI/Table/Table.vue (hunk 13 above)
Comment (dalmia) [reply src/components/UI/Table/Table.vue:219]: comment to explain what's going on

File: src/components/UI/Table/Table.vue [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+      const order = this.sortOrders[sortKey] || 1;
+
+      if (sortKey) {
+        data = data.slice().sort(function (a, b) {
+          a = a[sortKey];
+          b = b[sortKey];
+          if (sortKey == "name") {
+            a = a["title"].toLowerCase();
+            b = b["title"].toLowerCase();
+          }
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/components/UI/Table/Table.vue:251]: 1-comment to explain why

File: src/components/UI/Table/Table.vue [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+      // this is done so that the searching/sorting can be done on the plios as well
+      if (this.filteredData != undefined && this.filteredData[entryIndex] != undefined) {
+        this.filteredData[entryIndex]["name"] = {
+          ...this.filteredData[entryIndex]["name"],
+          ...plioDetails,
+        };
+      }
+    },
+    filterBySearch(data) {
+      const filterKey = this.filterKey && this.filterKey.toLowerCase();
```
[the remark this answers is not in the document]
Comment (deepansh96) [reply src/components/UI/Table/Table.vue:223]: https://stackoverflow.com/questions/17200315/logical-operator-and-two-strings-in-javascript/39025443#:~:text=4%20Answers&text=Because%20you're%20using%20%26%26,just%20the%20last%20evaluated%20thing.&text=If%20you%20just%20use%20a,see%20this%20output%20very%20easily.

File: src/components/UI/ListItems/PlioListItem.vue [hunk 16]
```diff
@@ -78,12 +78,8 @@ export default {
...
     plioId: {
       // the id of the plio to render
-      default: "",
-      type: String,
-    },
-    org: {
-      // the active workspace
-      default: "",
+      // default is "dummy"
+      default: "dummy",
```
Comment (dalmia) [inline src/components/UI/ListItems/PlioListItem.vue:82]: why can't we make use of the pending variable only instead of adding this new dummy thing?

File: src/services/API/Item.js [hunk 17]
```diff
@@ -20,8 +20,10 @@ export default {
     return apiClient().delete(itemsEndpoint + itemId);
   },
 
-  duplicateItem(itemId) {
-    // Create a clone of itemId
-    return apiClient().post(itemsEndpoint + itemId + "/duplicate");
+  duplicateItem(itemId, plioDBId) {
+    // Create a clone of itemId and link it to the prodided "plioDBId"
```
Comment (dalmia) [inline src/services/API/Item.js:24]:
```suggestion
    // Create a clone of itemId and link it to the provided plio's DB ID 
```

File: src/pages/Home.vue [hunk 18]
```diff
@@ -30,6 +28,11 @@ export default {
       searchQuery: "",
       tableColumns: ["name", "no. of learners"],
       tableData: [],
+      // dummy table data - to show skeletons when data is being loaded
+      dummyTableData: Array(5).fill({
+        name: { type: "component", value: "dummy" },
+        "no. of learners": "-",
+      }),
```
Comment (dalmia) [inline src/pages/Home.vue:35]: can we use the pending var or something else that already exists apart from creating a new dummy thing?

File: src/pages/Home.vue [hunk 19]
```diff
@@ -47,6 +51,7 @@ export default {
     },
 
     async prepareTableData(plioIdList) {
+      this.tableData = this.dummyTableData;
```
Comment (dalmia) [inline src/pages/Home.vue:54]: if we use the pending variable or something else, the dummy table values can be set within `Table.vue` and not Home.vue and then we can simply populate `this.tableData` without creating a new `tableData` var.

File: src/components/UI/Table/Table.vue [hunk 20]
```diff
@@ -55,9 +53,7 @@
                     :key="columnName"
                     scope="col"
                     class="sm:px-6 sm:py-3 px-3 py-1.5 text-left text-xs sm:text-md font-medium text-gray-500 uppercase tracking-wider"
-                    :class="{
-                      'hidden sm:table-cell': columnIndex != 0,
-                    }"
+                    :class="setColumnHeaderStyleClass(columnIndex)"
```
Comment (dalmia) [inline src/components/UI/Table/Table.vue:56]:
```suggestion
                    :class="getColumnHeaderStyleClass(columnIndex)"
```

File: src/components/UI/Table/Table.vue [hunk 21]
```diff
@@ -69,9 +65,7 @@
                         <inline-svg
                           :src="require('@/assets/images/chevron-down-solid.svg')"
                           class="h-3 w-3 my-1 transition ease duration-800"
-                          :class="{
-                            'transform rotate-180': sortOrders[columnName] == -1,
-                          }"
+                          :class="setSortIconStyleClass(columnName)"
```
Comment (dalmia) [inline src/components/UI/Table/Table.vue:68]:
```suggestion
                          :class="getSortIconStyleClass(columnName)"
```

File: src/components/UI/Table/Table.vue [hunk 22]
```diff
@@ -254,6 +270,32 @@ export default {
...
+        if (response.status == 201) {
+          this.$router.push({
+            name: "Editor",
+            params: { plioId: response.data.uuid, org: this.activeWorkspace },
+          });
+        } else {
+          this.$refs.toast.show("error", "Error creating Plio", this.toastLife);
+        }
+      });
+    },
```
Comment (dalmia) [inline src/components/UI/Table/Table.vue:298]: This should not be here - it should be in Home.vue

File: src/components/UI/Table/Table.vue [hunk 23]
```diff
@@ -114,11 +107,24 @@
...
+    <div
+      v-if="isTableEmpty"
+      class="bg-primary border border-2 shadow-lg rounded-2xl p-4 bg-white w-64 m-auto mt-32"
+    >
+      <icon-button
+        :titleConfig="createButtonTextConfig"
+        class="rounded-md mx-auto"
+        @click="createNewPlio"
+      ></icon-button>
+    </div>
```
Comment (dalmia) [inline src/components/UI/Table/Table.vue]: This should not be here - it should be in Home.vue

**Review by dalmia — CHANGES_REQUESTED** [review]

**Comment by dalmia** [comment]
- table should not be shown when no plios
- Button pe hover karne par cursor nahi change hota - button ke text par hover karne se hota hain - button ka hover pe color nahi change hota - for the create plio button
- Name column ka size is different for the same page width - this is not right na? same page width ke liye ek hi name ka width hona tha na?
- all plios shows 5 first and then changes to 18 - saara ek baar mein load hoke 18 dikhna chaiye na?

File: src/components/UI/ListItems/PlioListItem.vue (hunk 16 above)
Comment (deepansh96) [reply src/components/UI/ListItems/PlioListItem.vue:82]: will explain on call

**Comment by dalmia** [comment]
search and sorting should be disabled when skeleton

**Comment by dalmia** [comment]
Screenshot 2021-04-19 at 9 46 05 PM
- same screen two different button styles - looks weird

File: src/pages/Home.vue [hunk 24]
```diff
@@ -1,5 +1,12 @@
 <template>
   <Table :data="tableData" :columns="tableColumns" :tableTitle="'All Plios'"> </Table>
+  <!-- no plios exist warning -->
+  <div
+    v-if="showNoPliosCreatedWarning"
```
Comment (dalmia) [inline src/pages/Home.vue:5]:
```suggestion
    v-if="hasAnyPlios"
```

File: src/components/UI/ListItems/PlioListItem.vue [hunk 25]
```diff
@@ -138,10 +137,8 @@ export default {
   },
 
   async created() {
-    // load the plio only if the plio is not a dummy one
-    if (!this.isDummyPlio) {
-      await this.loadPlio();
-    }
+    // load the plio only if the plio id is not empty
+    if (this.plioId != "") await this.loadPlio();
```
Comment (dalmia) [inline src/components/UI/ListItems/PlioListItem.vue:141]: better to have this in a computed prop

**Review by dalmia — CHANGES_REQUESTED** [review]
Very minor changes requested.
